node/bft/src/primary.rs

```diff
@@ -916,9 +916,18 @@ impl<N: Network> Primary<N> {
         peer_ip: SocketAddr,
         certificate: BatchCertificate<N>,
     ) -> Result<()> {
+        // Ensure the batch certificate is from an authorized validator.
+        if !self.gateway.is_authorized_validator_ip(peer_ip) {
+            // Proceed to disconnect the validator.
+            self.gateway.disconnect(peer_ip);
+            bail!("Malicious peer - Received a batch certificate from an unauthorized validator IP ({peer_ip})");
+        }
         // Ensure storage does not already contain the certificate.
         if self.storage.contains_certificate(certificate.id()) {
             return Ok(());
+        // Otherwise, ensure ephemeral storage contains the certificate.
+        } else if !self.storage.contains_unprocessed_certificate(certificate.id()) {
+            self.storage.insert_unprocessed_certificate(certificate.clone())?;
```

Comment:
Why is the certificate not removed from the `unprocessed_certificates` cache when this function is completed?

Reply:
It is removed from the cache if the certificate is inserted into storage, so even sooner than waiting for the function to complete.

You might wonder whether a malicious validator could fill up the cache, and they could. The cache helps in the optimistic case to speed up processing.